plugins/org.jkiss.dbeaver.parser/src/org/jkiss/dbeaver/parser/GrammarAnalyzer.java

```diff
@@ -0,0 +1,188 @@
+/*
+ * DBeaver - Universal Database Manager
+ * Copyright (C) 2010-2022 DBeaver Corp and others
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jkiss.dbeaver.parser;
+
+import org.jkiss.dbeaver.parser.grammar.nfa.GrammarNfaBuilder.NfaFragment;
+import org.jkiss.dbeaver.parser.grammar.nfa.*;
+
+import java.util.*;
+
+/* 
+ * Class responsible for building terminal-guided automaton based on grammar graph
+ */
+class GrammarAnalyzer {
+
+    private final List<GrammarNfaTransition> terminalTransitions;
+    private final NfaFragment root;
+
+    /*
+     * Grammar graph traversal step
+     */
+    private static class Step implements Iterable<GrammarNfaTransition> {
+        private final Step prev;
+        private final boolean isUp;
+        private final GrammarNfaTransition transition;
+
+        private Step(Step prev, boolean isUp, GrammarNfaTransition transition) {
+            this.prev = prev;
+            this.isUp = isUp;
+            this.transition = transition;
+        }
+
+        public static Step initial(GrammarNfaTransition transition) {
+            return new Step(null, false, transition);
+        }
+
+        public Step enter(GrammarNfaTransition transition) {
+            return new Step(this, false, transition);
+        }
+
+        public Step exit(GrammarNfaTransition transition) {
+            return new Step(this, true, transition);
+        }
+
+        @Override
+        public String toString() {
+            ArrayList<String> steps = new ArrayList<>();
+            for (Step x = this; x != null; x = x.prev) {
+                steps.add(x.transition.toString());
+            }
+            return "Path of " + String.join(", ", steps);
+        }
+
+        @Override
+        public Iterator<GrammarNfaTransition> iterator() {
+            return new Iterator<GrammarNfaTransition>() {
+                private Step current = Step.this;
+
+                @Override
+                public GrammarNfaTransition next() {
```

Comment:
This implementation violates the general contract of this method (see javadoc for `Iterator.next()`).